sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +307,95 @@ def predict(self, X):
         """
         check_is_fitted(self)
 
-        X = self._validate_data(X, accept_sparse="csr", reset=False)
-        return self.classes_[
-            pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
-        ]
+        d = self._decision_function(X)
+        y_pred = self.classes_.take(d.argmax(1))
+        return y_pred
+
+    def _decision_function(self, X):
+        check_is_fitted(self, "centroids_")
+
+        X = self._validate_data(X, reset=False, accept_sparse="csr")
+
+        # check if this works with sparse matrix formats later
+        discriminant_score = np.empty(
+            (X.shape[0], self.classes_.size), dtype=np.float64
+        )
+
+        for cur_class in range(self.classes_.size):
+            Xdist = X - self.centroids_[cur_class, :]
+            Xdist_norm = np.square(Xdist / self.within_class_std_)
+            # Hastie et al. (2009), p. 652, Eq. (18.2)
+            discriminant_score[:, cur_class] = np.squeeze(
+                -np.sum(Xdist_norm, axis=1)
+                + 2.0 * np.log(self.class_priors_[cur_class])
+            )
+
+        return discriminant_score
+
+    def decision_function(self, X):
+        """Apply decision function to an array of samples.
+
+        The decision function is equal (up to a constant factor) to the
+        log-posterior of the model, i.e. `log p(y = k | x)`.
+        However, the decision function uses the argmax definition instead
+        of argmin definition as in the predict() function, which are just
+        -1 factor of each other.
+        The estimation has been implemented according to
+        Hastie et al. (2009), p. 652 equation (18.2)
```

Comment:
We need to be consistent with scikit-learn meaning that when thresholding the score, it should give us the expected class. So we need to be aligned with other classifier in this regards.

Reply:
And we don't need to document it explicitely.

Reply:
I don't understand. Can you give an example?

Reply:
Basically it comes back to check the sign of the `decision_function` in the binary classification case to be sure that we are consistent with other classifier.

With the definition above, and the current implementation in `decision_function`, the argmax seems to be the right way to go.